File: kaleido-gradle-plugin/src/main/java/com/tongsr/kaleido/gradle/R8ConfigurationEngine.java
package com.tongsr.kaleido.gradle;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeSet;

final class R8ConfigurationEngine {
    static final String SCHEMA = "KaleidoR8Configuration.v1";
    static final String PRODUCER = "KaleidoR8Configuration/1";
    static final int DICTIONARY_SIZE = 4096;

    private R8ConfigurationEngine() {}

    static Configuration generate(
            Map<String, String> adoption,
            ClassRewriteArtifacts.Plan classPlan) {
        var stream = required(adoption, "seed.domain.r8-dictionary");
        var applicationId = required(adoption, "applicationId");
        var plannedOutputs = Set.copyOf(classPlan.expectedOutputs());
        var fixedIdentities = new TreeSet<String>();
        for (var site : classPlan.manifestSites()) {
            var identity = resolveIdentity(site.target(), applicationId);
            if (!plannedOutputs.contains(identity)) {
                throw new IllegalArgumentException(
                        "Semantic class identity is absent from ClassRewritePlan outputs: "
                                + identity);
            }
            fixedIdentities.add(identity);
        }

        var dictionaries = new LinkedHashMap<String, String>();
        dictionaries.put("member.txt", dictionary(stream, "member", "m"));
        dictionaries.put("class.txt", dictionary(stream, "class", "C"));
        dictionaries.put("package.txt", dictionary(stream, "package", "p"));

        var rules = new StringBuilder("# ").append(SCHEMA).append('\n')
                .append("-obfuscationdictionary ../dictionaries/member.txt\n")
                .append("-classobfuscationdictionary ../dictionaries/class.txt\n")
                .append("-packageobfuscationdictionary ../dictionaries/package.txt\n");
        for (var identity : fixedIdentities) {
            rules.append("-keep,allowoptimization class ")
                    .append(identity).append(" { *; }\n");
        }
        return new Configuration(Map.copyOf(dictionaries), rules.toString(),
                List.copyOf(fixedIdentities));
    }

    private static String dictionary(String stream, String domain, String prefix) {
        var values = new TreeSet<String>();
        for (var index = 0; index < DICTIONARY_SIZE; index++) {
            var digest = SeedDerivation.derive(stream, "r8-" + domain, Integer.toString(index));
            var length = 10;
            var value = prefix + digest.substring(0, length);
            while (!values.add(value)) {
                length += 2;
                if (length > digest.length()) {
                    throw new IllegalStateException("Unable to allocate deterministic R8 token");
                }
                value = prefix + digest.substring(0, length);
            }
        }
        var text = new StringBuilder("# ").append(SCHEMA)
                .append(" kind=").append(domain).append('\n');
        values.forEach(value -> text.append(value).append('\n'));
        return text.toString();
    }

    private static String resolveIdentity(String lexical, String applicationId) {
        if (lexical.startsWith(".")) return applicationId + lexical;
        if (!lexical.contains(".")) return applicationId + "." + lexical;
        return lexical;
    }

    private static String required(Map<String, String> values, String key) {
        var value = values.get(key);
        if (value == null || value.isBlank()) {
            throw new IllegalArgumentException("Adoption Plan is missing " + key);
        }
        return value;
    }

    record Configuration(
            Map<String, String> dictionaries,
            String rules,
            List<String> fixedIdentities) {}
}

File: kaleido-gradle-plugin/src/main/kotlin/com/tongsr/kaleido/gradle/R8ConfigurationEngine.kt
package com.tongsr.kaleido.gradle

import java.util.LinkedHashMap
import java.util.TreeSet

internal object R8ConfigurationEngine {
    const val SCHEMA: String = "KaleidoR8Configuration.v1"
    const val PRODUCER: String = "KaleidoR8Configuration/1"
    const val DICTIONARY_SIZE: Int = 4096

    @JvmStatic
    fun generate(
        adoption: Map<String, String>,
        classPlan: ClassRewriteArtifacts.Plan,
    ): Configuration {
        val stream = required(adoption, "seed.domain.r8-dictionary")
        val applicationId = required(adoption, "applicationId")
        val plannedOutputs = classPlan.expectedOutputs.orEmpty().toSet()
        val fixedIdentities = TreeSet<String>()
        for (site in classPlan.manifestSites.orEmpty()) {
            val identity = resolveIdentity(requireNotNull(site?.target), applicationId)
            if (identity !in plannedOutputs) {
                throw IllegalArgumentException(
                    "Semantic class identity is absent from ClassRewritePlan outputs: $identity",
                )
            }
            fixedIdentities.add(identity)
        }

        val dictionaries = LinkedHashMap<String, String>()
        dictionaries["member.txt"] = dictionary(stream, "member", "m")
        dictionaries["class.txt"] = dictionary(stream, "class", "C")
        dictionaries["package.txt"] = dictionary(stream, "package", "p")

        val rules = StringBuilder("# ").append(SCHEMA).append('\n')
            .append("-obfuscationdictionary ../dictionaries/member.txt\n")
            .append("-classobfuscationdictionary ../dictionaries/class.txt\n")
            .append("-packageobfuscationdictionary ../dictionaries/package.txt\n")
        for (identity in fixedIdentities) {
            rules.append("-keep,allowoptimization class ")
                .append(identity).append(" { *; }\n")
        }
        return Configuration(dictionaries.toMap(), rules.toString(), fixedIdentities.toList())
    }

    private fun dictionary(stream: String, domain: String, prefix: String): String {
        val values = TreeSet<String>()
        for (index in 0 until DICTIONARY_SIZE) {
            val digest = SeedDerivation.derive(stream, "r8-$domain", index.toString())
            var length = 10
            var value = prefix + digest.substring(0, length)
            while (!values.add(value)) {
                length += 2
                if (length > digest.length) {
                    throw IllegalStateException("Unable to allocate deterministic R8 token")
                }
                value = prefix + digest.substring(0, length)
            }
        }
        val text = StringBuilder("# ").append(SCHEMA)
            .append(" kind=").append(domain).append('\n')
        values.forEach { value -> text.append(value).append('\n') }
        return text.toString()
    }

    private fun resolveIdentity(lexical: String, applicationId: String): String {
        if (lexical.startsWith(".")) return applicationId + lexical
        if (!lexical.contains(".")) return "$applicationId.$lexical"
        return lexical
    }

    private fun required(values: Map<String, String>, key: String): String {
        val value = values[key]
        if (value.isNullOrBlank()) {
            throw IllegalArgumentException("Adoption Plan is missing $key")
        }
        return value
    }

    @JvmRecord
    data class Configuration(
        val dictionaries: Map<String, String>,
        val rules: String,
        val fixedIdentities: List<String>,
    )
}
